Multiplayer campaign

MrChuse opened the pull request:
![image](https://github.com/freegish/freegish/assets/43815531/1c6ad35d-3c10-4efd-8832-fbb3c3cee3c9)
I wanted to play the campaign with my friends and I added a picker for the number of players to the main menu which is activated by pressing the Y button (like Turbo mode is activated using T button)

I also fixed an annoying behavior which duplicated videomodes in video settings, now I can finally access my 1920x1080p
Before:
![image](https://github.com/freegish/freegish/assets/43815531/a0246721-0b0a-4c7b-8364-75b984e565dc)
After:
![image](https://github.com/freegish/freegish/assets/43815531/04b11127-01ad-48f6-b09b-44fd0535fcae)

And I renamed a lot of hardcoded variables with values from an enum located in gameobject.h (maybe there's a better way to structure these, but I like having all the constants in the same file)

If the multiplayer campaign is not needed but you want any of the other two changes, I can create another pull request from another branch.

MrChuse committed:
Implemented multiplayer campaign with camera zoomout

MrChuse committed:
Removed duplicate resolutions

MrChuse committed:
Started with object types in setup.c

MrChuse committed:
Changed object_types to level_object_types and added gameobject_types
for gameobject.c

MrChuse committed:
Added rope_types for gameobject.c

MrChuse committed:
Added available types to objfunc.c

MrChuse committed:
Added IS_ACTIVATED, IS_ONE_TIME to buttons and switches and IS_SECRET to areaswitches

MrChuse committed:
Added light_type

MrChuse committed:
Changed more hardcoded variables to enum values

MrChuse committed:
Merge pull request #1 from MrChuse/readability

Readability

MrChuse committed:
Merge pull request #2 from MrChuse/resolution-fix

Removed duplicate resolutions

MrChuse committed:
Merge branch 'master' into multiplayer-campaign

MrChuse committed:
Merge pull request #3 from MrChuse/multiplayer-campaign

Implemented multiplayer campaign with camera zoomout

MrChuse commented:
I did not test the Hera fight and it might be broken but when we'll reach this level, I'll write back

MrChuse committed:
Fix to make heads spawn on their corresponding bodies

MrChuse committed:
Fixed LVL_OBJ_TYPE vs OBJ_TYPE mistake causing wheels to fold over themselves

MrChuse committed:
Fixed misplaced ropes and links on Hera's level

MrChuse committed:
Newly created Heras will attack their own player now instead of grouping on the first player and ignoring others

MrChuse commented:
> I did not test the Hera fight and it might be broken but when we'll reach this level, I'll write back

Alright, done with the Hera level I guess, official campaign is totally playable now (with two players, didn't test the spawns of 3 and 4)
One minor point is that all the mobs attack the first player but its so minor that I don't want to fix it right now, maybe later when I have free time

Asmageddon commented:
@MrChuse Hmm, y'know, maybe I could just give you permissions to the repo? Honestly none of the  maintainers have touched the repo in like a decade, and I don't even code much nowadays :')

@FrozenCow, @alvinx, @bobvanderlinden, what do you think of this idea?

bobvanderlinden commented:
Sounds good to me. I haven't done anything with Gish in a long time, so would be great if someone wants to continue development!

MrChuse commented:
> @MrChuse Hmm, y'know, maybe I could just give you permissions to the repo? Honestly none of the maintainers have touched the repo in like a decade, and I don't even code much nowadays :')
> 
> @FrozenCow, @alvinx, @bobvanderlinden, what do you think of this idea?

I just want someone to actually look at and possibly test my changes.
I'm fine with working in either my fork or this upstream repo, if there are no maintainers left, maybe I could become one.
What does a maintainer do?

MrChuse committed:
Fixed exception in listfiles function

MrChuse committed:
Fixed savelevel and loadlevel operate in different folders

MrChuse committed:
Reverted savelevel to version 10

MrChuse committed:
Merge pull request #4 from MrChuse/editor-fix

Editor fix

MrChuse committed:
Moved panning view in editor(s) into a separate function, added speed up when shift is held

MrChuse committed:
Added more entries to level object types enum and changed editor to show their name instead of type id

MrChuse committed:
Reordered text in the object editor

MrChuse committed:
Added picking from a list of gametypes

MrChuse committed:
Now the editor opens without need to press F1 first

MrChuse committed:
Added single level button to the player menu

MrChuse committed:
Moved zoom to a separate function too

MrChuse committed:
Added saving current rope type and showing what the type is in ropeeditor

MrChuse committed:
Removed setting unlocks when the game is loaded

MrChuse committed:
Renamed unlocks

MrChuse committed:
Renamed difficulties too

MrChuse committed:
Added debug option to look at player's unlocks

MrChuse committed:
Added pipette tool to editor

MrChuse committed:
Merge pull request #5 from MrChuse/editor-fix

Editor fix

MrChuse committed:
Fixed version 10 load and readded version 11 save

MrChuse committed:
Fixed a bug where textures that load on create new level don't match with those after save and load

MrChuse committed:
Started to rename hardcoded textures in the texture array

MrChuse committed:
Single level button now only appears after a game is beaten on any difficulty

MrChuse committed:
Unsuccessful try to bind existing texturefiles to textures loaded from ver10 and ver9 levelfiles

MrChuse committed:
Implemented texture binding

MrChuse committed:
Fixed resetting filename when copying textures from tilesets in textureedit

MrChuse committed:
Moved adding a single animation to a function

MrChuse committed:
Moved adding an animation part to a function

MrChuse committed:
Moved loading head part to a function

MrChuse committed:
Merge pull request #6 from MrChuse/editor-fix

Editor fix

MrChuse committed:
Added attacking the closest player

MrChuse committed:
Merge pull request #8 from MrChuse/Attack-closest-player

Added attacking the closest player

MrChuse committed:
Merge pull request #7 from MrChuse/development

Added multiple editor fixed and various code refactor improvements

MrChuse commented:
Alright, I added more useful stuff. As for the main topic, the enemies now attack the closest player to them (except enemy Heras, they attack their corresponding player). However, I added many more improvements to various files. My favorites are [d8ce73f](https://github.com/freegish/freegish/pull/25/commits/d8ce73fd21e99d5e137d884e56196b7a36730d19) [92ad8c9](https://github.com/freegish/freegish/pull/25/commits/92ad8c9f10a78c160c73713c4f41a0db61cf2501) and [b523277](https://github.com/freegish/freegish/pull/25/commits/b5232776c38ea60c3e2bd4c00fdb0ecd56153373).
Also some improvements to the editor such as naming the object and rope you are about to place and a pipette tool (press E to copy). A bug where saving and loading a level would mismatch the textures was also fixed.

MrChuse committed:
Reconnecting gamepad works now

MrChuse committed:
Fixed crashes caused by modifying string literals

MrChuse committed:
Merge pull request #9 from MrChuse/load-joystick

Reconnecting gamepad works now

Asmageddon commented:
@MrChuse Well, as the maintainer, you'd have access to this repository to push to, and it'd be your responsibility to ensure it compiles, etc. You could also reach out to whoever manages any potential builds in repositories - e.g. [freegish-git in AUR](https://aur.archlinux.org/packages/freegish-git).

As for looking over your code, I'm sorry but I'd pretty much have to wrap my head around a codebase I only ever briefly looked at, and I struggle with chronic fatigue and brainfog nowadays, so that's a tall order. I'd rather add you to the freegish org so you can work on it yourself :P

MrChuse commented:
@Asmageddon ok, I don't see any reason to not become a maintainer, let's do it

MrChuse committed:
Fixed variable length in the loadanimation

Asmageddon commented:
@MrChuse Alright I sent you an invite to the freegish organisation.

MrChuse committed:
Merge branch 'development'

MrChuse committed:
Factored out keybinding rows in options menu into a separate function

MrChuse committed:
Refactored duplication of drawkeypickoption into a loop

MrChuse committed:
Refactored code into functions to bind keys (keyboard and joysticks). Disabled ability to erase keys from the same and other players when already bound key was pressed\.

MrChuse committed:
Overhaul of keybinds using a preset system. Integrated preset system
with versus mode and multiplayer campaign.

MrChuse committed:
Added deletion of presets. Removed optionsmenu2, using presets now. Fixed a bug of not deselecting preset name input

MrChuse committed:
Merge pull request #10 from MrChuse/keybinding-profiles

Keybinding profiles

MrChuse committed:
Added more debugs and fixed remembering old incorrect texture paths

MrChuse committed:
Merge pull request #11 from MrChuse/keybinding-profiles

Annoying editor fix

MrChuse committed:
Fixed disability to leave 4-player party

MrChuse committed:
Added a cute screenshot to title

MrChuse committed:
Fixed a bug with drawtext exception

MrChuse committed:
Added a tile01 folder to have at least something in the editor, pure white is so eye burning.

MrChuse committed:
Added a funny picture to storymode.png

MrChuse committed:
Added campaign thoughts

MrChuse committed:
Removed stupid chdir, loading non-existing level files now does not crash further loadings

MrChuse committed:
Created vsmode.png with screenshots from gamemodes

MrChuse committed:
Added textures for all versus gamemodes

MrChuse committed:
Merge pull request #12 from MrChuse/drawing-overhaul

Drawing overhaul

MrChuse committed:
Get rid of prompting selecting players each time in versus mode. Now the game asks the first time the versus mode is clicked and then it remembers it.

MrChuse committed:
Added a player select when a profile is selected instead of defaulting to keyboard profile 1

MrChuse committed:
Merge pull request #13 from MrChuse/micro-fixes

Micro fixes

MrChuse committed:
Moved files to datapacks directory

MrChuse committed:
Added datapacks variables and loading textures from these folders

MrChuse committed:
loading music from datapacks too

MrChuse committed:
loading sounds from datapacks too

MrChuse committed:
loading levels from datapacks too

MrChuse committed:
loading animations from datapacks too

MrChuse committed:
renamed function load_all_textures to load_all_text_textures to remove collisions

MrChuse committed:
Removed forgotten chdir in loadanimation. Animations are always loaded now (removed check)

MrChuse committed:
Moved loading animations to the main menu because lazy loading is hard to pair up with datapacks

MrChuse committed:
Fixed saving levels working with datapacks. Copied setting filename to 0 in loading version 9 like in version 10

MrChuse committed:
Fixed setup_presets_to_players definition

MrChuse committed:
Added names to menu items in editor

MrChuse committed:
Added ability to change datapacks to options menu. Moved all appropriate loading code to a separate load_datapack function

MrChuse committed:
Added official datapack to gitignore

MrChuse committed:
Moved font, mouse and wav sounds to load_datapack too

MrChuse committed:
Merge pull request #14 from MrChuse/datapacks

Datapacks

MrChuse committed:
Bind presets to controls only in relevant places: after loading the player menu and after exiting setup screen

MrChuse committed:
Restarting in the single level menu does not send us back to the menu now.

MrChuse committed:
Added reset_level menu variable

MrChuse committed:
Added debug options to the options menu

MrChuse committed:
Added debug_render_bonds to debug options. Works in the gameloop

MrChuse committed:
Added render player vertices and render level lines to debug options

MrChuse committed:
Changed debug character positions to debug objectnums and objectlinks in appropriate places

MrChuse committed:
Merge pull request #15 from MrChuse/debug-options

Debug options

MrChuse commented:
Hello @Asmageddon, I've made some more changes and would like them merged. Can you send me a maintainer invitation once more, please?

henriquesimoes committed:
Use standard boolean type

In C23, bool is a new reserved keyword [1], which means that a custom
enum cannot be named after it anymore. Replace it with a proper bool
type, which fulfills the original purpose of this enum.

Because not all compilers support C23 yet, including MSVC and Apple
Clang [2], include the compatibility header stdbool.h available since
C99 so it builds fine in all platforms.

This change is especially relevant because GCC 15 defaults to C23,
causing freegish to fail to build [3].

[1] https://gcc.gnu.org/gcc-15/porting_to.html#c23-new-keywords
[2] https://en.cppreference.com/w/c/compiler_support/23
[3] https://bugs.debian.org/1096653

MrChuse committed:
Merge pull request #16 from henriquesimoes/c23

Pulling commit with compatibility with C++

MrChuse committed:
Added range of required versions from 2.9 to 4.1

MrChuse committed:
Added build file to streamline building

MrChuse committed:
Fixed vcpkg command

MrChuse committed:
Fixed GAMEXIT typo (GAMEEXIT now)

MrChuse committed:
Added GAMEOVER enum, changed campaign code to use it

MrChuse committed:
Used gameover enum in other gametypes

MrChuse committed:
Added levelarea typedef

MrChuse committed:
Reworded NONE checks, added GAMEOVER_WIN_GOOD and GAMEOVER_WIN_BAD to make it clear

MrChuse committed:
Merge pull request #17 from MrChuse/cmake

Added range of required versions from 2.9 to 4.1

MrChuse committed:
Merge pull request #18 from MrChuse/enums

Enums

MrChuse committed:
Deleted backup.lvl level file

MrChuse committed:
Fixed gametype picker

MrChuse committed:
Editor reopens last level when closing and opening editor without exiting the whole game

MrChuse committed:
Texture filename clarity

MrChuse committed:
Overhauled sewer1.lvl to contain all the mechanics and tutorial lines

MrChuse committed:
Overhauled sewer2.lvl, wrote some notes on the first chapter of original Gish

MrChuse committed:
Added reports to loading a level in the editor, cleared by mouseclick

MrChuse committed:
Fixed objedit pipette

MrChuse committed:
Added P key to set gish position in editor

MrChuse committed:
Added M key to Move objects around. Mass field hotkey was changes to S

MrChuse committed:
Added loading texture in editblock

MrChuse committed:
Added sewer3, added boxes to sewer1, modified thoughts accordingly

MrChuse committed:
Changed Mass hotkey back to M. Who knew that S was for moving down, huh. Hotkey to Move objects was changed to P (Place)

MrChuse committed:
Added ability to remove players with mouse and delete key, I had a joystick preset stuck :(

MrChuse committed:
Overhauled sewer4.lvl

MrChuse committed:
Overhauled sewer5.lvl

MrChuse committed:
Moved entrance to the left in sewer6

MrChuse committed:
Added lighting to levels

MrChuse committed:
Added coins to levels

MrChuse committed:
Minor changes to sewer1

MrChuse committed:
Added V key to paste settings from one object to another. Doesn't copy ropes

MrChuse committed:
Overhauled sewer7

MrChuse committed:
Changed brightness to 1 in all sewer levels

MrChuse committed:
Added LVL_OBJ_TYPE for all the mobs

MrChuse committed:
Added sample textures for sewer mobs

MrChuse committed:
Added a second jump to the jump tutorial in sewer1

MrChuse committed:
Reworded tutorial climb text

MrChuse committed:
Fixed vs modes links and debug info

MrChuse committed:
Fixed jump tutorial

MrChuse committed:
Fixed Joystick load crash & Added logs

MrChuse committed:
Changes after playtest

MrChuse committed:
Changed jump tutorial text

MrChuse committed:
Added water texture

MrChuse committed:
Added playroom, it has only jumps for now

MrChuse committed:
Added menuinputmouse to be able to check for mouse menu actions

MrChuse committed:
Added setup, load, save buttons to the editor

MrChuse committed:
Added Show lines and Focus layer buttons to editor

MrChuse committed:
Second world thoughts

MrChuse committed:
Added buttons for picking tile layer

MrChuse committed:
Merged editlevelobject with editlevel. F6 is the binding to edit blocks

MrChuse committed:
Merged ropeedit into the editlevel function

MrChuse committed:
Buttons for Tiles, Objects and Ropes are now visualized

MrChuse committed:
Merge pull request #20 from MrChuse/editor

Editor

MrChuse committed:
Merge pull request #21 from MrChuse/editor-all-in-one

Editor all in one

MrChuse committed:
Fixed Rope type name index typo

MrChuse committed:
Fixed loading audio: use correct folder and stereo wav files

MrChuse committed:
Merge pull request #22 from MrChuse/fix-sound

Fixed loading audio: use correct folder and stereo wav files
